Fix few baseline calculation cases for tables with empty rows

Ahmad-S792 opened the pull request:
#### 28d5064d279331b68e2fe8487cb317c271fc1299
<pre>
Fix few baseline calculation cases for tables with empty rows

<a href="https://bugs.webkit.org/show_bug.cgi?id=285159">https://bugs.webkit.org/show_bug.cgi?id=285159</a>
<a href="https://rdar.apple.com/142046863">rdar://142046863</a>

Reviewed by Alan Baradlay.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Partial Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/9fa80d5986647e8be2f1b19fc5e722cb51da4e27">https://chromium.googlesource.com/chromium/src.git/+/9fa80d5986647e8be2f1b19fc5e722cb51da4e27</a>

CSS says: &quot;If a row has no cell box aligned to its baseline, the baseline of that
row is the bottom content edge of the lowest cell in the row.&quot;

This partial merge progresses following cases of baseline in WPT test case:

&gt; Empty row&apos;s baseline is center of the row.
&gt; First row defines baseline even if empty.
&gt; Two rows empty, 1st row has css height.
&gt; First row empty with css height, 2 row has content

* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::firstLineBaseline const):
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/tentative/baseline-table-expected.txt: Rebaselined
* LayoutTests/platform/glib/fast/dynamic/insert-before-table-part-in-continuation-expected.txt:
* LayoutTests/platform/ios/fast/dynamic/insert-before-table-part-in-continuation-expected.txt:
* LayoutTests/platform/mac/fast/dynamic/insert-before-table-part-in-continuation-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-tables/tentative/baseline-table-expected.txt: Rebaselined

Canonical link: <a href="https://commits.webkit.org/288314@main">https://commits.webkit.org/288314@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f7406180029361ad47469dfcb201bcab12511e2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82440 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2104 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87572 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33501 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84545 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2175 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9992 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/64256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/22009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85509 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/1549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/75007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/44533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/1447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/29193 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32542 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/29828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88929 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9746 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7007 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9972 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/70821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/71874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/16014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1122 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12798 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9699 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15220 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9573 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13039 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11343 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1593,7 +1593,11 @@ std::optional<LayoutUnit> RenderTable::firstLineBaseline() const
         return std::optional<LayoutUnit>(topNonEmptySection->logicalTop() + baseline.value());
 
-    // FIXME: A table row always has a baseline per CSS 2.1. Will this return the right value?
+    // FF, Presto and IE use the top of the section as the baseline if its first row is empty of cells or content.
```
both Presto (Opera) and Trident (IE) have disappeared. Maybe the reference to these is not useful. We could simply say "Other browsers".

Ahmad-S792 replied:
Yep - I kept it to make it easy in future to find any relevant Blink merges but will update once it get r+ before merging.

Ahmad-S792 replied:
Done. :-)

Ahmad-S792 committed:
Fix few baseline calculation cases for tables with empty rows

https://bugs.webkit.org/show_bug.cgi?id=285159
rdar://142046863

Reviewed by Alan Baradlay.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Partial Merge: https://chromium.googlesource.com/chromium/src.git/+/9fa80d5986647e8be2f1b19fc5e722cb51da4e27

CSS says: "If a row has no cell box aligned to its baseline, the baseline of that
row is the bottom content edge of the lowest cell in the row."

This partial merge progresses following cases of baseline in WPT test case:

> Empty row's baseline is center of the row.
> First row defines baseline even if empty.
> Two rows empty, 1st row has css height.
> First row empty with css height, 2 row has content

* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::firstLineBaseline const):
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/tentative/baseline-table-expected.txt: Rebaselined
* LayoutTests/platform/glib/fast/dynamic/insert-before-table-part-in-continuation-expected.txt:
* LayoutTests/platform/ios/fast/dynamic/insert-before-table-part-in-continuation-expected.txt:
* LayoutTests/platform/mac/fast/dynamic/insert-before-table-part-in-continuation-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-tables/tentative/baseline-table-expected.txt: Rebaselined

Canonical link: https://commits.webkit.org/288314@main
